File: ios/native_download_manager/Sources/native_download_manager/NativeDownloadManagerPlugin.swift
import Flutter
import UIKit

public class NativeDownloadManagerPlugin: NSObject, FlutterPlugin, NativeDownloadManagerHostApi {
  
  private var manager = DownloadManager.shared
  
  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = NativeDownloadManagerPlugin()
    
    // Set up Pigeon Host API
    NativeDownloadManagerHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    
    // Set up Pigeon Flutter API
    let flutterApi = NativeDownloadManagerFlutterApi(binaryMessenger: registrar.messenger())
    DownloadManager.shared.flutterApi = flutterApi
    
    registrar.addMethodCallDelegate(instance, channel: FlutterMethodChannel(name: "native_download_manager_dummy", binaryMessenger: registrar.messenger()))
  }
  
  // NativeDownloadManagerHostApi protocols
  func initialize() {
    // Explicit initialization hook
  }
  
  func startDownload(request: PigeonDownloadRequest) {
    manager.startDownload(request: request)
  }
  
  func pauseDownload(taskId: String) {
    manager.pauseDownload(taskId: taskId)
  }
  
  func resumeDownload(taskId: String) {
    manager.resumeDownload(taskId: taskId)
  }
  
  func cancelDownload(taskId: String) {
    manager.cancelDownload(taskId: taskId)
  }
  
  func retryDownload(taskId: String) {
    manager.retryDownload(taskId: taskId)
  }
  
  func deleteDownload(taskId: String, deleteFile: Bool) {
    manager.deleteDownload(taskId: taskId, deleteFile: deleteFile)
  }
  
  func getAllTasks() -> [PigeonDownloadTask] {
    return manager.getAllTasks()
  }
  
  func getTask(taskId: String) -> PigeonDownloadTask? {
    return manager.getTask(taskId: taskId)
  }
  
  func clearHistory() {
    manager.clearHistory()
  }
  
  func setConcurrencyLimit(limit: Int64) {
    // iOS manages concurrent downloads automatically on background URLSession, 
    // but we can set HTTPMaximumConnectionsPerHost on a configuration if needed.
  }
}

File: android/src/main/kotlin/com/native_download_manager/NativeDownloadManagerPlugin.kt
package com.native_download_manager

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin

class NativeDownloadManagerPlugin : FlutterPlugin, NativeDownloadManagerHostApi {
    private var context: Context? = null
    private var controller: DownloadManagerController? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        controller = DownloadManagerController.getInstance(flutterPluginBinding.applicationContext)
        
        // Setup Pigeon Host API
        NativeDownloadManagerHostApi.setUp(flutterPluginBinding.binaryMessenger, this)
        
        // Setup Pigeon Flutter API for callbacks
        val flutterApi = NativeDownloadManagerFlutterApi(flutterPluginBinding.binaryMessenger)
        controller?.flutterApi = flutterApi
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        NativeDownloadManagerHostApi.setUp(binding.binaryMessenger, null)
        controller?.flutterApi = null
        controller = null
        context = null
    }

    // NativeDownloadManagerHostApi implementations
    override fun initialize() {
        // Initialization is done implicitly on attached, but this acts as an explicit hook.
    }

    override fun startDownload(request: PigeonDownloadRequest) {
        controller?.enqueue(request)
    }

    override fun pauseDownload(taskId: String) {
        controller?.pause(taskId)
    }

    override fun resumeDownload(taskId: String) {
        controller?.resume(taskId)
    }

    override fun cancelDownload(taskId: String) {
        controller?.cancel(taskId)
    }

    override fun retryDownload(taskId: String) {
        controller?.retry(taskId)
    }

    override fun deleteDownload(taskId: String, deleteFile: Boolean) {
        controller?.delete(taskId, deleteFile)
    }

    override fun getAllTasks(): List<PigeonDownloadTask> {
        return controller?.getAllTasks() ?: emptyList()
    }

    override fun getTask(taskId: String): PigeonDownloadTask? {
        return controller?.getTask(taskId)
    }

    override fun clearHistory() {
        controller?.clearHistory()
    }

    override fun setConcurrencyLimit(limit: Long) {
        controller?.setConcurrencyLimit(limit.toInt())
    }
}
